images/airflow/2.9.2/python/mwaa/celery/task_monitor.py

```diff
@@ -273,15 +286,48 @@ def _cleanup_undead_process(process_id: int):
     )
 
 
+def _get_next_unprocessed_signal() -> (str, dict):
+    signal_search_start_timestamp = math.ceil((datetime.now(tz=tz.tzutc()) - SIGNAL_SEARCH_TIME_RANGE).timestamp())
+    signal_filenames = os.listdir(MWAA_SIGNALS_DIRECTORY) if os.path.exists(MWAA_SIGNALS_DIRECTORY) else []
+    sorted_filenames = sorted(signal_filenames)
+    for signal_filename in sorted_filenames:
+        # Signals file follow the naming convention "<createTimestamp>_<signalId>.json"
+        file_timestamp = signal_filename.split("_")[0]
+        # In case of issues, signals can arrive late and out of order. So, we scan all unprocessed signals in a search time range.
+        # Only a handful of signals are expected for each worker, so this repeated processing should be very light.
+        if file_timestamp.isdigit() and int(file_timestamp) > signal_search_start_timestamp:
+            signal_file_path = os.path.join(MWAA_SIGNALS_DIRECTORY, signal_filename)
+            with open(signal_file_path, "r") as file_data:
+                signal_data = json.load(file_data)
+                if not signal_data["processed"]:
+                    return signal_file_path, signal_data
+    return None, None
+
+
+def _marked_signal_as_processed(signal_filepath, signal_data):
+    signal_data["processed"] = True
```

Comment:
Following from above, if this user has write access do we have the risk of customers modifying/deleting the signals?  What is the behavior in that case?

Reply:
These files are created by the serviceEni container. The serviceEni container will look into just the data relevant to the corresponding Fargate task which means that these files will be specific to the worker processing them. 

If the customer for some reason modify these files, then they can make the task monitor make wrong decisions regarding when the worker should terminate or not. But none of the service side state is affected by it and we will have timeouts in place at the service-side to force correct this.